src/Compiler/RefC/RefC.idr

```diff
@@ -363,75 +355,90 @@ emit fc line = do
         (Yes _) => flip snoc (lJust indentedLine maxLineLengthForComment ' ' ++ " " ++ comment)
         (No _)  => flip appendR [indentedLine, (lJust ""   maxLineLengthForComment ' ' ++ " " ++ comment)]
 
-
-freeTmpVars : {auto t : Ref TemporaryVariableTracker (List (List String))}
-           -> {auto oft : Ref OutfileText Output}
-           -> {auto il : Ref IndentLevel Nat}
-           -> Core $ ()
-freeTmpVars = do
-    lists <- get TemporaryVariableTracker
-    case lists of
-        (vars :: varss) => do
-            traverse_ (\v => emit EmptyFC $ "removeReference(" ++ v ++ ");" ) vars
-            put TemporaryVariableTracker varss
-        [] => pure ()
-
-
 addHeader : {auto h : Ref HeaderFiles (SortedSet String)}
          -> String
          -> Core ()
 addHeader = update HeaderFiles . insert
 
+applyFunctionToVars : {auto oft : Ref OutfileText Output}
+                    -> {auto il : Ref IndentLevel Nat}
+                    -> String
+                    -> List String
+                    -> Core $ ()
+applyFunctionToVars fun vars = traverse_ (\v => emit EmptyFC $ fun ++ "(" ++ v ++ ");" ) vars
+
+removeVars : {auto oft : Ref OutfileText Output}
+           -> {auto il : Ref IndentLevel Nat}
+           -> List String
+           -> Core $ ()
+removeVars = applyFunctionToVars "removeReference"
+
+dupVars : {auto oft : Ref OutfileText Output}
+           -> {auto il : Ref IndentLevel Nat}
+           -> List String
+           -> Core $ ()
+dupVars = applyFunctionToVars "newReference"
+
+removeReuseConstructors : {auto oft : Ref OutfileText Output}
+                        -> {auto il : Ref IndentLevel Nat}
+                        -> List String
+                        -> Core $ ()
+removeReuseConstructors = applyFunctionToVars "removeReuseConstructor"
 
+avarToC : Env -> AVar -> Core String
+avarToC (MkPercEnv borrowed owned _) x = do
+    pure $
+        if contains x borrowed then "newReference(" ++ varName x ++ ")"
+        else if contains x owned then varName x
+        else assert_total $ idris_crash "INTERNAL ERROR: variable " ++ varName x ++ " is not owned and borrowed "
 
 makeArglist : {auto a : Ref ArgCounter Nat}
-           -> {auto t : Ref TemporaryVariableTracker (List (List String))}
            -> {auto oft : Ref OutfileText Output}
            -> {auto il : Ref IndentLevel Nat}
+           -> Env
```

Comment:
Do you mean something like `{auto t : Ref EnvTracker (List Env)}`? In that case I'm not sure what to do in `avarToC` function. What if the list `List Env` is empty? I think passing the environment in the argument explicitly makes the flow of control clearer. Is it really important for us to use `Ref` instead?

Reply:
Where does the list come from? Why not just `Env`?

Reply:
If you only ever need to access the latest one you could define
```idris
locally : {auto t : Ref EnvTracker Env} -> Env -> Core a -> Core a
locally newEnv act
  = do oldEnv <- get EnvTracker
       put EnvTracker newEnv
       act
       put EnvTracker oldEnv
```

Reply:
I literally meant that you could declare `locally` and then use it as a combinator
precisely so that there is never any risk to forget to call the last `put`.